File: apps/api/app/schemas/history.py
from datetime import datetime
from enum import StrEnum
from typing import Any
from uuid import UUID

from pydantic import BaseModel, ConfigDict


class ActivityEventType(StrEnum):
    INCOME_CREATED = "income_created"
    INCOME_UPDATED = "income_updated"
    INCOME_DELETED = "income_deleted"
    EXPENSE_CREATED = "expense_created"
    EXPENSE_UPDATED = "expense_updated"
    EXPENSE_DELETED = "expense_deleted"
    CATEGORY_CREATED = "category_created"
    CATEGORY_UPDATED = "category_updated"
    CATEGORY_ARCHIVED = "category_archived"
    FILE_UPLOADED = "file_uploaded"
    FILE_DELETED = "file_deleted"
    EXTRACTION_STARTED = "extraction_started"
    EXTRACTION_COMPLETED = "extraction_completed"
    EXTRACTION_FAILED = "extraction_failed"
    AI_DRAFT_CONFIRMED = "ai_draft_confirmed"
    MEMBER_ADDED = "member_added"
    MEMBER_REMOVED = "member_removed"
    ROLE_CHANGED = "role_changed"
    SETTING_CHANGED = "setting_changed"


class ActivityHistoryItem(BaseModel):
    id: UUID
    event_type: ActivityEventType
    actor_user_id: UUID | None = None
    actor_display_name: str | None = None
    entity_table: str
    entity_id: UUID | None = None
    summary: dict[str, Any]
    created_at: datetime

    model_config = ConfigDict(from_attributes=True)


class HistoryPage(BaseModel):
    items: list[ActivityHistoryItem]
    next_before: datetime | None = None

File: apps/web/lib/api/history.ts
import { apiFetch } from "./client";

export type ActivityEventType =
  | "income_created"
  | "income_updated"
  | "income_deleted"
  | "expense_created"
  | "expense_updated"
  | "expense_deleted"
  | "category_created"
  | "category_updated"
  | "category_archived"
  | "file_uploaded"
  | "file_deleted"
  | "extraction_started"
  | "extraction_completed"
  | "extraction_failed"
  | "ai_draft_confirmed"
  | "member_added"
  | "member_removed"
  | "role_changed"
  | "setting_changed";

export type ActivityHistoryItem = {
  id: string;
  event_type: ActivityEventType;
  actor_user_id: string | null;
  actor_display_name: string | null;
  entity_table: string;
  entity_id: string | null;
  summary: Record<string, unknown>;
  created_at: string;
};

export type HistoryPage = {
  items: ActivityHistoryItem[];
  next_before: string | null;
};

export async function listHistory(
  workspaceId: string,
  options: { limit?: number; before?: string | null } = {},
) {
  const params = new URLSearchParams();
  params.set("limit", String(options.limit ?? 50));
  if (options.before) {
    params.set("before", options.before);
  }

  return apiFetch<HistoryPage>(`/workspaces/${workspaceId}/history?${params.toString()}`);
}
